### PR TITLE
Add debug callback param to pass manager

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit adds a new optional parameter to the PassManager method for
enabling debug callback function. When this is set PassManager.run() will call the
callback function with the name, output dag, and pass number for each pass that the
 pass manager runs This can be useful when trying to debug why/when a particular
pass manager setup is going wrong, and provides users with the flexibility to inspect
and interact with the output of each pass as it runs (one example use case is to visualize
the output of each pass).
stage as it's executed.

### Details and comments

### CODE REVIEW

**Comment by nonhermitian** [comment]
What about just a callback that returns the circ or dag?  Then let the user decide what to do with it. But I like this.

**Comment by 1ucian0** [comment]
How is this PR different from #2450 ?

**Comment by mtreinish** [comment]
@nonhermitian sure, I can make it a callback that will pass the circuit and pass name to whatever function gets passed in. That'll give people more flexibility to analyze the output of each pass as they need.

@1ucian0 #2450 adds logging about the execution of each pass when it starts, ended, and the elapsed time to execute each pass. But it doesn't actually show what each stage did to the circuit, which can be useful for debugging. That's what this patch adds.

**Comment by ajavadia** [comment]
I agree both this and #2450 are useful. 

But I think the API should be the same. Currently, this debugging is done via an arg of `pass_manager.run()`, while the logging in #2450 is done via a setter `pass_manager.log_passes = True`. The output of this gets printed (or callback), while the output of #2450 goes in `pass_manager.property_set`. Can we agree on one?

Also if you're exposing the flag at `transpile_circuit()`, should probably expose it at `transpile()` as well.

**Comment by 1ucian0** [comment]
If the user provides her own `PassManagerContext`, that can be used for hooking into then before-after-each-pass time. And the time logging from #2450 is a particular instance of `PassManagerContext`. What do you think?

**Comment by nonhermitian** [comment]
So I still think a callback is the best.  This way the user can do whatever they want.  I would also argue that the passmanager works on DAGs and most useful info besides visualization can be found from the DAG itself.  So I think that that would be the appropriate return object.  I would not want to waste time on the conversion unless the user syas so themselves.  One must also be careful about parallel executions with a callback turned on.  Maybe one needs to return the index associated with a given circuit in the input list in this case.  Something like

```python

def callback(dag, idx):
    do stuff here...

pm = PassManager(..., callback=callback)

transpile(circs, backend, passmanager=pm)
```

**Comment by 1ucian0** [comment]
With a context, you can do things before and after the pass execution. Almost as you described, but
```python
class MyPassContext(PassContext):
    def __enter__(self):
        # do stuff before running the pass

    def __exit__(self, *exc_info):
        # do stuff after running the pass

pm = PassManager(..., context=MyPassContext)
transpile(circs, backend, passmanager=pm)
```

This would allow us to the `pass_manager.log_passes = True` (which is a bit ugly) and unify with the interface wit the timer logger.

**Comment by nonhermitian** [comment]
This just looks like a more complicated version of a callback to me. Providing the dag to users via a simple callback is inline with what many other packages do when the intermediate results of computations are desired (e.g. see SciPy).  And there, as in most other places, the callback is called 'callback'.  If we return the name of the pass as well, then most information that is not in the scope of the passmanager (e.g. the relationship between passes) is available to users however they want it.

**Comment by ajavadia** [comment]
Just pointing out that the Callback and PassManagerContext will force you to use a PassManager instance. It will not, for example, work with `transpile(circuit, FakeTenerife(), optimization_level=2)`. @mtreinish's debug flag does work with that as well, but I think the Callback is probably the more standard way to do this. @1ucian0 what is an example of something the PassManagerContext can handle which the Callback cannot?

**Comment by nonhermitian** [comment]
Yes, but is it too much to ask that one grab an instance of O2 pm to do a debugging?  I think if you can debug a pm you can grab a pm instance and pass it to the transpiler.

**Comment by mtreinish** [comment]
I agree with Paul here, the interface he's proposing seems like the best tradeoff for flexibility and simplicity. It also isn't that hard to manually provision a preset pass manager and just set pm.callback before passing it to transpile. I've updated this PR to use that approach instead of printing to stdout as before.

**Comment by 1ucian0** [comment]
@ajavadia said:
> @1ucian0 what is an example of something the PassManagerContext can handle which the Callback cannot?

The timer logger in #2450

**Comment by 1ucian0** [comment]
@mtreinish said:
> It also isn't that hard to manually instantiate a preset pass manager and just set pm.callback before passing it to transpile.

With the context also could work in the same way.

**Comment by nonhermitian** [comment]
Actually you can, provided that the call back returns the original unaltered dag first. Then the only diff would be the small overhead from calling the backend func. However this time is well within the variance of the pass timing.

**Comment by 1ucian0** [comment]
Other advantage that I see with the context-approach is flexibility in the parameters.  The callback-approach is fixed in the parameters `def callback(dag, idx)`. If in the future we want to extend  the available objects to the call back, for example, pass instance, we have to modify the existing callbacks to `def callback(dag, idx, pass_)` (or tell our users to write callbacks as `def callback(dag, **kwargs)`). In the context-approach, we just need to extend the `__init__` in `PassContext`, which is under our control.

**Comment by 1ucian0** [comment]
> Actually you can, provided that the call back returns the original unaltered dag first. Then the only diff would be the small overhead from calling the backend func. However this time is well within the variance of the pass timing.

It's more complicated. Take:
```
pm = PassManager(passA, callback=callback)
``` 
In the current #2579 implementation, the function `callback` runs at the end of `passA._do_pass`. But that might be more than a single transformation. If passA requires passB, then `callback` is called only ones and there is no way to hook in between.

In the recursive way we resolve requirements, i don't see a trivial way to implement timer log. But code is welcome.

**Comment by nonhermitian** [comment]
Actually it is not possible.  Not because of any inherent information, but because you should only return a dag from transformation passes, as those are the only ones that modify the dag, and would have any meaning for debugging from a dag.

One could also argue, and I would agree, that the timing is in the context of the pass manager, ie. it is doing the timing, and a context would make sense in that case. For example, one would time each individual pass, but also the aggregate time of a loop that iterates until convergence.  So this would fall under the scope of the pass manager which is supposed to control passes and their relationships.  In contrast the dag is never a property of the pass manager; it is an input and output only.  In that case a callback make sense.  Again, this how many numerical algorithms give access to internal states.

**Comment by 1ucian0** [comment]
I tried to implement time logger with the approach in #2579 and I'm getting very different results. I'm opened to callback solution if there is a way to implement timer log. Based on the test in #2579, here is my attempt:
```python
...
calls = {} 

def callback(*args): 
     calls[args[1]] =  "%.5f (ms)" % ((time()-calls['start']) * 1000) 
     calls['start'] = time() 

passmanager = PassManager(callback=callback)
...
passmanager.log_passes = True 
calls['start'] = time() 
transpile(circuit, FakeRueschlikon(), pass_manager=passmanager)
```
```python
> passmanager.property_set['pass_log']                                                                                                             
['Unroller: 2.25496 (ms)', 'Optimize1qGates: 2.54989 (ms)']
```
```python
> calls                                                                                                                                            
{'start': 1559919578.694869,
 'Unroller': '53.47705 (ms)',
 'Optimize1qGates': '4.78697 (ms)'}
```

How can be improved?

**Comment by 1ucian0** [comment]
Now the question is if we want to have a non-unified way to handle callback and timer logger. Are you okey with that @ajavadia?

----
> you should only return a dag from transformation passes, as those are the only ones that modify the dag, and would have any meaning for debugging from a dag.

I'm not fully sure if I agree with debugging only makes sense for transformation passes. But we can leave this discussion for another time.

**Comment by 1ucian0** [comment]
Just for the record, unifying the interfaces means that the callback code would live in `MyPassContext.__exit__`:

```python
class MyPassContext(PassContext):
    def __exit__(self, *exc_info):
        print(self.dag)
```
vs
```python
def callback(dag):
     print(dag)
```
I understand is more complicated, but I think is worthy.

**Comment by ajavadia** [comment]
@nonhermitian:
> Yes, but is it too much to ask that one grab an instance of O2 pm to do a debugging?
No. I am fine with debugging only being available from a PassManager instance. I was just pointing out the difference.

@1ucian0:
> I tried to implement time logger with the approach in #2579 and I'm getting very different results.
What is the source of this difference? I would prefer a unified approach to debugging that gives you both handles to the internal state as well as timing info. But I recognize that it may not be easy to achieve.

**Comment by 1ucian0** [comment]
> What is the source of this difference?

Because it measuring more than the `pass.run` time. The most precise way to measure is starting the stopwatch before and finish it after. That requires hooking into those two points.

> I would prefer a unified approach to debugging that gives you both handles to the internal state as well as timing info. But I recognize that it may not be easy to achieve.

I also would prefer to have a unified approach.

**Comment by 1ucian0** [comment]
The `PassContext` suggestion is available in #2601.

**Comment by kdk** [comment]
Late to the discussion, but I'll add my two cents. 

1) When run in parallel, any callback/context manager code will be run somewhere in a `multiprocessing.Pool`. We should make sure it's possible to get the results of any visualization or analysis back to the user, or otherwise only allow the case when `len(circuits) == 1`

2) I'm not a fan that both the callback and context manager approaches require the user to know, in advance, what questions they want to ask (and how to ask them). Debugging would involve an iterative loop of write a callback/context manager, transpile, inspect outcome, update callback/context manager, repeat.

For e.g. optimization, where data structures may be large and the number of iterations may be huge, this makes sense. But for transpilation, neither of these are true. Why not just return a rich data structure including all the information we would provided to the callback/context manager, and let the user interact with it directly?

The user has to write no additional code (just set a keyword argument) and has immediate access to all of the relevant metadata/debugging information.

**Comment by nonhermitian** [comment]
Part 1: See https://github.com/Qiskit/qiskit-terra/pull/2579#issuecomment-499518706

Part 2: I am not sure how one goes and debugs something if the following is not true:
> require the user to know, in advance, what questions they want to ask (and how to ask them).

Indeed, you need to first have an idea that something is wrong, e.g. the depth is not what you guessed, to many gates, etc.  Then you need to go and find out where things are going wrong by checking the quantities you think are incorrect.  If you can not do this, then how can you debug something?

> Why not just return a rich data structure including all the information we would provided to the callback/context manager, and let the user interact with it directly?

Because the user has to learn yet another object, and this does not follow the conventions used for iterative algorithms.  In these cases it is a simple callback that returns the data structure being iterated over.   If the SciPy example is not enough to convince you, then feel free to look at tensorflow as well.  Also, callbacks are implemented in aqua, so a simple callback provides for a more cohesive interface over the elements.

> For e.g. optimization, where data structures may be large and the number of iterations may be huge, this makes sense.

I thought the transpiler is an optimizer.  Indeed, we call it a circuit optimization workflow here: https://github.com/Qiskit/qiskit-tutorials/blob/master/qiskit/terra/using_the_transpiler.ipynb, and have several different `optimization_level` passmanagers predefined.  

Also, can you guarantee that the number of iterations and data structures will continue to be small as the width and depth of the circuits that we can run grows?

> The user has to write no additional code (just set a keyword argument) and has immediate access to all of the relevant metadata/debugging information.

How does one retrieve the rich data structure without adding additional code?  Presumably one must change the transpiler calls to something like:

```python
circs, rich_data_structure = transpile(......, return_rich_structure=True)
```

**Comment by mtreinish** [comment]
> Late to the discussion, but I'll add my two cents.
> 
>     1. When run in parallel, any callback/context manager code will be run somewhere in a `multiprocessing.Pool`. We should make sure it's possible to get the results of any visualization or analysis back to the user, or otherwise only allow the case when `len(circuits) == 1`

It should work fine, but a user will have to write a callback function so that the output expects that multiple circuits will be run in parallel. You can easily just pull the process id and append that to whatever your output is. FWIW, also from the passmanager class itself it doesn't look there is anything about being part of a pool with multiple circuits being run in parallel, it happens at the layer above. That being said typing this made me realize we should return the circuit name along with the other information in the callback just so the extra information is there for people if they don't want to use a process or thread id to differentiate between parallel executions. I'll add that to this shortly.

> 
>     2. I'm not a fan that both the callback and context manager approaches require the user to know, in advance, what questions they want to ask (and how to ask them). Debugging would involve an iterative loop of write a callback/context manager, transpile, inspect outcome, update callback/context manager, repeat.
> 
> 
> For e.g. optimization, where data structures may be large and the number of iterations may be huge, this makes sense. But for transpilation, neither of these are true. Why not just return a rich data structure including all the information we would provided to the callback/context manager, and let the user interact with it directly?
> 
> The user has to write no additional code (just set a keyword argument) and has immediate access to all of the relevant metadata/debugging information.

This was my original idea (albeit in a flawed/incomplete way) in this PR by just having the passmanager print the circuit diagram for each pass, which obviously isn't enough. The problem is outside of writing to something global (like stdout, which has a lot of headaches in the multiprocess part you mentioned earlier) there isn't really a way to go about this. Especially, if we're going to store all the relevant info for debugging that we want people to interact with for debugging, there isn't really a way to extract that info from the call. That's why having a callback function that takes that data as its input made sense to me and I respun this PR. It provides an interface to extract the data for debugging during execution of the passmanager without changing the function signatures to `transpile()` or `passmanager.run()` so they can access information needed to debug without too much modification to the code path being debugged.

**Comment by kdk** [comment]
> Part 1: See [#2579 (comment)](https://github.com/Qiskit/qiskit-terra/pull/2579#issuecomment-499518706)

I'm not sure I see how this solves the problem. Won't we still hit an error when we try to pickle the callback to send it out into the pool (assuming it would be defined locally)? And won't some side effects of the callback behave differently if they're executed in the main interpreter vs in the pool?

> 
> Part 2: I am not sure how one goes and debugs something if the following is not true:
> 
> > require the user to know, in advance, what questions they want to ask (and how to ask them).
> 

You should see me in `pdb`. Debugging is an exploitative process. If every time the user has a new question, a new idea, they have to update their code and re-transpile, it becomes that much more burdensome.

> 
> > Why not just return a rich data structure including all the information we would provided to the callback/context manager, and let the user interact with it directly?
> 
> Because the user has to learn yet another object, and this does not follow the conventions used for iterative algorithms. In these cases it is a simple callback that returns the data structure being iterated over. If the SciPy example is not enough to convince you, then feel free to look at tensorflow as well. Also, callbacks are implemented in aqua, so a simple callback provides for a more cohesive interface over the elements.

Not necessarily, it doesn't have to be another object, but certainly it's harder to embrace the repl from within a callback. If I have an object, I can print it, put it in a dataframe, graph it, realize I have the wrong datatypes, cast and regraph it, and I can do this only having had to transpile the circuits once. 

> 
> > For e.g. optimization, where data structures may be large and the number of iterations may be huge, this makes sense.
> 
> I thought the transpiler is an optimizer. Indeed, we call it a circuit optimization workflow here: https://github.com/Qiskit/qiskit-tutorials/blob/master/qiskit/terra/using_the_transpiler.ipynb, and have several different `optimization_level` passmanagers predefined.
> 
> Also, can you guarantee that the number of iterations and data structures will continue to be small as the width and depth of the circuits that we can run grows?
> 

No, and I wouldn't expect it to. Nor am I suggesting that these approaches are mutually exclusive. Tensorflow, AFAICT, has methods to do both. I'm suggesting we should build first toward the situation we face now, when most of our circuits and optimizations are not above that threshold. 

> > The user has to write no additional code (just set a keyword argument) and has immediate access to all of the relevant metadata/debugging information.
> 
> How does one retrieve the rich data structure without adding additional code? Presumably one must change the transpiler calls to something like:
> 
> ```python
> circs, rich_data_structure = transpile(......, return_rich_structure=True)
> ```

```python
circs, debug_data = transpile(......, debug=True)
print(debug_data)
```
One extra line  :). I'd argue that learning the signature of the callback/context manager and writing it is at least as hard, if not harder than `print(debug_data)`.

**Comment by kdk** [comment]
> > Late to the discussion, but I'll add my two cents.
> > ```
> > 1. When run in parallel, any callback/context manager code will be run somewhere in a `multiprocessing.Pool`. We should make sure it's possible to get the results of any visualization or analysis back to the user, or otherwise only allow the case when `len(circuits) == 1`
> > ```
> 
> It should work fine, but a user will have to write a callback function so that the output expects that multiple circuits will be run in parallel. You can easily just pull the process id and append that to whatever your output is. FWIW, also from the passmanager class itself it doesn't look there is anything about being part of a pool with multiple circuits being run in parallel, it happens at the layer above. That being said typing this made me realize we should return the circuit name along with the other information in the callback just so the extra information is there for people if they don't want to use a process or thread id to differentiate between parallel executions. I'll add that to this shortly.

I agree it's possible, but it imposes constrains on what the user can do within the callback. e.g. How hard would it be to graph the number of swaps added by say `StochasticSwap` across a set of 100 different circuits? (As a side point, maybe a view into the output state of the PropertySet be a useful thing to have?)

> 
> > ```
> > 2. I'm not a fan that both the callback and context manager approaches require the user to know, in advance, what questions they want to ask (and how to ask them). Debugging would involve an iterative loop of write a callback/context manager, transpile, inspect outcome, update callback/context manager, repeat.
> > ```
> > 
> > 
> > For e.g. optimization, where data structures may be large and the number of iterations may be huge, this makes sense. But for transpilation, neither of these are true. Why not just return a rich data structure including all the information we would provided to the callback/context manager, and let the user interact with it directly?
> > The user has to write no additional code (just set a keyword argument) and has immediate access to all of the relevant metadata/debugging information.
> 
> This was my original idea (albeit in a flawed/incomplete way) in this PR by just having the passmanager print the circuit diagram for each pass, which obviously isn't enough. The problem is outside of writing to something global (like stdout, which has a lot of headaches in the multiprocess part you mentioned earlier) there isn't really a way to go about this. Especially, if we're going to store all the relevant info for debugging that we want people to interact with for debugging, there isn't really a way to extract that info from the call. That's why having a callback function that takes that data as its input made sense to me and I respun this PR. It provides an interface to extract the data for debugging during execution of the passmanager without changing the function signatures to `transpile()` or `passmanager.run()` so they can access information needed to debug without too much modification to the code path being debugged.

:+1: I do agree that the callback minimizes the modification to the user's code, and the pass manager, and see the benefit there. I'm becoming more of the opinion that these two are complimentary approaches. If there's agreement that there's value in both, I can draw up a PR to collect and return debug data.

**Comment by kdk** [comment]
After talking some with @1ucian0 , I'm more of the opinion that the best approach here is to expose either or both of the the callback and debug_data interfaces, but implement them internally as individual `PassContext`s.  This would give us the ability to present a familiar interface for debugging and allow us to consolidate the logging and debugging code in the pass manager. Thoughts?

**Comment by ajavadia** [comment]
There are a few ideas floating here that I'll try to summarize and figure out how they work with an example use case. Thanks to @kdk who walked me through this as I was quite confused.

**Usecase**: Say I transpile 100 circuits, and want to check how long the "Unroller" pass took on average, and then I want to draw the 3rd circuit after it has gone through the Unroller pass.

## 1. callback

```python
# I propose this signature (note the dag already has the circuit name to distinguish processes)
def callback(pass_, output_dag, time, property_set):
   if isinstance(pass_, Unroller):
       unroller_time.append(time)
       print('avg unroller time: ', 
              sum(unroller_time)/len(unroller_time))
   if output_dag.name == circuits[2].name:
     print(dag_to_circuit(output_dag))

pm = PassManager(callback=callback)
circuits = pm.run(circuits)  # this should work too: circuits = transpile(circuits, callback=callback)
```

*pro*:
- It is very flexible, you can even code inject and alter the path of the transpilation (e.g. change the layout in the middle).
- It gives you incremental data. That would be useful for situations where the transpiler errors or does not terminate.

*con*:
- It is not completely trivial to guess that the callback has to have that particular signature, must consult documentation. 

vs. 

## 2. debug=True flag
Setting a flag returns a rich data of what happened during the transpilation, which can be queried etc.
```python
pm = PassManager(debug=True)
circuits, debug_data = pm.run(circuits)  # also: circuit, debug_data = transpile(circuit, debug=True)

for circ in debug_data:
    for pass_ in circ:
         if isinstance(pass_, Unroller):
             unroller_time.append(pass_['time'])
             print('avg unroller time: ', 
                   sum(unroller_time)/len(unroller_time))
             if pass_['dag'].name == circuits[2].name:
                   print(dag_to_circuit(pass_['dag']))
```

```python
# debug data would be a list of lists (outer one circuits, inner ones passes). Each list element is a dict, pertaining to information about the pass.

print(debug_data)

[
 # circuit 0
  [{'pass': 'Unroller'
     'time': 44 (ms)
     'dag': <DAGCircuit>
      'property_set': ...
     },  # pass 0
    {'pass': 'CXCancellation'
     'time': 66 (ms)
     'dag': <DAGCircuit>
      'property_set': ...
     }],  # pass 1

  # circuit 1
  [{'pass': 'CommutativeCancellation'
     'time': 78 (ms)
     'dag': <DAGCircuit>
      'property_set': ...
     }]   

....
]
```

*pro:*
- You get the dump of all the debug data once, then you can keep asking new questions by querying that datastructure. Whereas in the callback approach, you have to change your callback function and add the question you are asking to it, and *rerun the transpiler*.

*con:*
- It's not obvious how to navigate this datastructure (list of list of dicts)
- The return signature changes when you set a flag.

## 3. Logging
I think we need to have logging, but use Python's logging features with levels etc. This should just be something you turn on and get pretty-printed outputs written to the console. Can be used for simple debugging, but not as sophisticated as the above two.

## 4. PassContext (implemented in https://github.com/Qiskit/qiskit-terra/pull/2601)
I don't really understand what this adds. It is not timing, because timing can be done with the callback and debug flag as well (see examples above). And it looks quite complicated to use. 


@nonhermitian @mtreinish do you agree with the signature of the above callback?

@1ucian0 can you rework the time log that you did to pretty-print timings via the Python logger, and to save raw timing in the callback approach above?

**Comment by mtreinish** [comment]
@ajavadia I'm fine with that signature, I'll defer to you on what the necessary data to pass back is. I'll modify the PR shortly to change the signature. In addition i'll add a `callback` kwarg to `transpile()`.

I had a couple of questions though, if the circuit name is in the intermediate dag objects the whole time  why does `run()` method bother storing the name and reassigning it to the final output circuit? The other thing I'm unclear about (also mostly because of unfamiliarity with the internals of the pass manager) is whether the property set will let someone writing a callback ascertain the position of a particular pass in the pass manager pipeline (which is what we were using count for before). I think there is value in returning that information somehow, otherwise we're forcing users to build that on their own (likely by adding it to an object outside the callback function scope).

**Comment by 1ucian0** [comment]
> 4. PassContext (implemented in #2601)
> I don't really understand what this adds. It is not timing, because timing can be done with the
> callback and debug flag as well (see examples above). And it looks quite complicated to use.

The approach in #2601 allows to implement a time logger as an instance of a pass contexts. The callback approach use that data, but the time logging was done independently.

**Comment by mtreinish** [comment]
@ajavadia  ok, I've reworked the signature to align more with what you were suggesting. I left the count argument (we can rename it to index or something if that's more descriptive) too so we don't require using a outer scope structure to recreate the position when callback is called. Also to address concerns about changing this interface over time I switched to using kwargs so that we can add new values without worrying about breaking people with existing callback functions.

File: qiskit/transpiler/passmanager.py [hunk 1]
```diff
@@ -39,7 +41,37 @@ def __init__(self, passes=None,
                 None.
             max_iteration (int): The schedule looping iterates until the condition is met or until
                 max_iteration is reached.
+            callback (func): A callback function that will be called after each
+                pass execution. The function will be called with 4 keyword
```
Comment (ajavadia) [inline qiskit/transpiler/passmanager.py]:
```suggestion
                pass execution. The function will be called with 5 keyword
```

File: qiskit/transpiler/passmanager.py [hunk 2]
```diff
@@ -133,9 +165,23 @@ def run(self, circuit):
...
         self.reset()  # Reset passmanager instance before starting
 
+        count = 0
         for passset in self.working_list:
             for pass_ in passset:
+                if self.callback:
+                    start_time = time()
                 dag = self._do_pass(pass_, dag, passset.options)
+                if self.callback:
+                    end_time = time()
```
Comment (ajavadia) [inline qiskit/transpiler/passmanager.py]: timing the `_do_pass` call won't be accurate because this function does a pass as well as any requirements of that pass, which may be other passes. 

I think it has to use the context around the call to `_run_this_pass()` which @1ucian0 suggested (line 221 of this file)

File: qiskit/transpiler/passmanager.py [hunk 3]
```diff
@@ -133,9 +165,23 @@ def run(self, circuit):
...
+                if self.callback:
+                    run_time = end_time - start_time
+                    # Deepcopy dag to ensure reference doesn't get updated
+                    # if callback func uses the dag after the loop iteration
+                    # continues
+                    self.callback(pass_=pass_, dag=copy.deepcopy(dag),
```
Comment (ajavadia) [inline qiskit/transpiler/passmanager.py]: I would say one of the advantages of the callback is that people can "inject code", meaning get a handle to an intermediate dag, modify it, and continue. A kind of interactive debugging. So I don't think we should prevent that here. It's the user's responsibility what to do with the dag.

Comment (mtreinish) [reply qiskit/transpiler/passmanager.py]: Sure, I did the deepcopy mostly for the tests, but I agree it's probably better to just do the deepcopy in the test's callback function instead of here.

File: qiskit/transpiler/passmanager.py (hunk 2 above)
Comment (mtreinish) [reply qiskit/transpiler/passmanager.py]: Right, I saw how the time was being done there, but it raised larger questions about how we not just measure time but also when we call the callback and what the expectation is for what's passed to it. If we get the time like where the context manager does now that can result in multiple times being generated per iteration in the `passmanager.run()` loop, but with only 1 callback function call. So the question is what do we consider a pass for the callback function is it the full pass (including the execution of any requirements) or each inidivual Pass object that we call `run()` on. I did time this way under the assumption it was the former and we were going to treat the complete execution of `_do_pass()` as what the callback deals with, and used that for the timing. If we want it to be each call of run on a pass object I can change that, but we can't really do it both ways.

**Comment by mtreinish** [comment]
One other thing I found while testing the transpile() call is that dag.name is not always defined for the output dag from each pass (which answers my earlier question about why circuit name is saved and later restored outside the run loop). I'm thinking that I'll have to add circuit_name back because of this so we can at least have a way to identify which circuit callback calls are for in the parallel multiple circuits case.

File: qiskit/transpiler/passmanager.py [hunk 4]
```diff
@@ -39,7 +40,37 @@ def __init__(self, passes=None,
...
+                forward. But, additional arguments may be added in future
+                release. To ensure your callback functions work against any
+                version of terra moving forward you want to have your callback
+                function take a single `**kwargs` input.
+
+                To use the callback feature you define a function that will
+                take in kwargs dict and access the variables. For example::
+
+                    def callback_func(**kwargs):
+                        pass = kwargs['pass']
```
Comment (ajavadia) [inline qiskit/transpiler/passmanager.py]:
```suggestion
                        pass_ = kwargs['pass_']
```

File: qiskit/compiler/transpile.py [hunk 5]
```diff
@@ -120,6 +120,36 @@ def transpile(circuits,
...
+            forward. But, additional arguments may be added in future
+            release. To ensure your callback functions work against any
+            version of terra moving forward you want to have your callback
+            function take a single `**kwargs` input.
+
+            To use the callback feature you define a function that will
+            take in kwargs dict and access the variables. For example::
+
+                def callback_func(**kwargs):
+                    pass = kwargs['pass']
```
Comment (ajavadia) [inline qiskit/compiler/transpile.py]:
```suggestion
                    pass_ = kwargs['pass_']
```

File: qiskit/transpiler/passmanager.py (hunk 2 above)
Comment (ajavadia) [reply qiskit/transpiler/passmanager.py]: I think the debugger should explicitly show all the passes that were run, because that is useful information that can help debugging. For example say I create a pass manager with a single pass like this and debug it during run:

```python
from qiskit.transpiler import PassManager
from qiskit.transpiler.passes import CommutativeCancellation

def callback(pass_, dag, time, property_set, count):
     print(pass_)

pm = PassManager()
pm.append(CommutativeCancellation())
pm.callback = callback

circ = QuantumCircuit(3)
circ.z(0)
circ.cx(0, 2)
circ.z(0)

pm.run(circ)
```
It will show exactly one pass as being run, even though really two passes are run (`CommutationAnalysis`, then, `CommutiveCancellation`), because of the `requires` relationship between the passes which the PassManager fulfills.

File: qiskit/compiler/transpile.py [hunk 6]
```diff
@@ -120,6 +120,36 @@ def transpile(circuits,
...
+                time (float): the time to execute the pass
+                property_set (PropertySet): the property set
+                count (int): the index for the pass execution
+
+            Keyword arguments are used for forward compatibility reasons,
+            and the current options will always be available moving
+            forward. But, additional arguments may be added in future
+            release. To ensure your callback functions work against any
+            version of terra moving forward you want to have your callback
+            function take a single `**kwargs` input.
```
Comment (ajavadia) [inline qiskit/compiler/transpile.py]: I'm not sure backwards compatibility is so important here, and something we should promise at this stage. This is just a debugging tool that should be used sporadically, not something to write notebooks with, etc.
For example if we merge DAGCircuit with QuantumCircuit, then the `dag` kwarg will be lost.

Comment (mtreinish) [reply qiskit/compiler/transpile.py]: Sure, I'll reword this to instead warn there is no interface guarantee moving forward.

File: test/python/transpiler/test_passmanager.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        expected_end = QuantumCircuit(qr)
+        expected_end.u2(0, np.pi, qr[0])
+        expected_end_dag = circuit_to_dag(expected_end)
+
+        calls = []
+
+        def callback(**kwargs):
+            out_dict = kwargs
+            out_dict['dag'] = copy.deepcopy(kwargs['dag'])
+            calls.append(kwargs)
```
Comment (itoko) [inline test/python/transpiler/test_passmanager.py]: Maybe `calls.append(out_dict)`?

Comment (mtreinish) [reply test/python/transpiler/test_passmanager.py]: oops, yeah good catch, thanks. Fixed in: https://github.com/Qiskit/qiskit-terra/pull/2579/commits/2222bc18409afb21ea5b8d6e7e8e36a3e2d02866

File: qiskit/transpiler/passmanager.py (hunk 2 above)
Comment (mtreinish) [reply qiskit/transpiler/passmanager.py]: Ok I reworked this to call the callback right after each `pass_.run()` call. To do it cleanly required removing the use of the context manager because it got tricky to extract the time measurements from it and adding the callback to the context manager made it overly complex.

**Comment by ajavadia** [comment]
This is looking good. The `PassManager.callback` is a superset of `PassManager.log_passes` now, so I propose removing the other one, and do proper logging instead (within Python's logging framework). I'll open a separate issue for that.
